include/circt/Dialect/Handshake/HandshakeOps.td

```diff
@@ -1249,3 +1249,138 @@ def EndOp : Handshake_Op<"end", [Terminator]> {
     mlir::ValueRange getMemoryControls();
   }];
 }
+
+//===----------------------------------------------------------------------===//
+// Speculation operations
+//===----------------------------------------------------------------------===//
+
+def SpeculatorOp : Handshake_Op<"speculator", [ 
+  TypesMatchWith<"data operand type matches data result type",
+                 "dataIn", "dataOut", "$_self">]> {
+  let summary = "Central control unit of the speculative circuit.";
+  let description = [{
+    The speculator produces speculative tokens when real data has
+    not yet arrived. The speculator is also responsible for deciding 
+    the correctness of its speculation.
+
+    It needs to send control tokens to save units and commit units
+    ($saveCtrl, $commitCtrl). For save-commit units, three control
+    signals are needed. $SCCtrl is connected directly, while 
+    $SCCtrlOptional replicates the branches that $dataOut
+    follows and might not reach a save-commit. $SCBranchCtrl is needed
+    in the replicated branches to discard the optional signal.
+
+    Example:
+    ```mlir
+    %dataOut, %saveCtrl, %commitCtrl, %SCCtrl, 
+      %SCCtrlOptional, %SCBranchCtrl = speculator %dataIn : i1
```

Comment:
I really don't like 'optional'- What you intended here was 'conditional' which is not quite the same thing.

I'd be curious if there was a good way to phrase "save-like" and "commit-like" control signals? Ctrl1 is used to control the saving/reading behavior of the save-commit and Ctrl2 is used to control some of the killing, the commit bit of the save-commit. But it might be too much information to try encode in a variable name.

Reply:
Perhaps a custom MLIR type instead of a simple `i3` would be clearer (never done anything like this myself but certainly possible)? Or custom getters on the operation itself to get a specific bit of the i3 operands by "purpose"? Fine by me if we declare this to be "future improvement material", just wanted to mention our possibilities.

Reply:
We're constrained to [this kind of a structure](https://sunset-hardware-3ee.notion.site/8ab7584e10bd4e48b0e4f61c2c81891e?pvs=4) until we change the VHDL code, and I think 3 bits with one 'standard' control signal and one conditional control signal does convey that structure the best.  

The VHDL code needs quite a big overhaul, but getting what we had before running with MLIR is the priority before looking at it, so it occasionally forces us into making slightly weird short term decisions.

Reply:
I have changed the names of the save-commit control signals as @murphe67 suggested.